### PR TITLE
Added measures

### PR DESCRIPTION
**Pull request by Paul1298**
But not SymmetricUncertainty because I don't understand what it does

### CODE REVIEW

File: filters/Filter.py [hunk 1]
```diff
@@ -1,18 +1,92 @@
 from functools import partial
-
+from sklearn.feature_selection import mutual_info_classif as MI
+from sklearn.metrics import mutual_info_score as MI_features
```
Comment (LastShekel) [inline filters/Filter.py]: We are not using sklearn at all, should write mutual information by yourself

File: filters/Filter.py [hunk 2]
```diff
@@ -1,18 +1,92 @@
...
+            # Main calculations
+            for sample_index, value in enumerate(feature):  # For each sample value
+                for i in range(tokensN):  # For each class token
+                    # Here can be raise warnings by 0/0 division. In this case, default results
+                    # are interpreted correctly
+                    distances[i] = np.abs(value - centers[i]) / variances[i]
+                fc[feature_index] += np.argmin(distances) == y[sample_index]
+
+        fc /= y.shape[0]
+        return fc
```
Comment (LastShekel) [inline filters/Filter.py]: Measure function should return dictionaries like {feature name: measure}
Example {"feature 1":0.5,"feature 5":0.0}

File: filters/Filter.py [hunk 3]
```diff
@@ -1,18 +1,92 @@
...
+            return f_ratio
+
+        f_ratios = []
+        for feature in X.T:
+            f_ratio = __calculate_F_ratio__(feature, y.T)
+            f_ratios.append(f_ratio)
+        f_ratios = np.array(f_ratios)
+        # return top n f_ratios TODO maybe add n into parameters?
+        n = 10
+        return np.argpartition(f_ratios, -n)[-n:]
```
Comment (LastShekel) [inline filters/Filter.py]: same as above

File: filters/Filter.py [hunk 4]
```diff
@@ -26,15 +100,148 @@ def gini_index(X, y):
...
+                        dict_y[y[i]] += 1
+
+                # calculate the probability of corresponding label.
+                sub_entropy = 0.0
+                for l in dict_y.values():
+                    sub_entropy += -l / sum(dict_y.values()) * math.log(l / sum(dict_y.values()), 2)
+
+                con_entropy += sub_entropy * p
+            list_f[index] = entropy - con_entropy
+        return list_f
```
Comment (LastShekel) [inline filters/Filter.py]: same as above

File: filters/Filter.py [hunk 5]
```diff
@@ -26,15 +100,148 @@ def gini_index(X, y):
...
+
+    # return list(index)
+    # i'm used MID as default. Or we need add ability to choose info_gain?
+    @staticmethod
+    def mrmr_measure(X, y):
+        def _find_first_feature(X, y):
```
Comment (LastShekel) [inline filters/Filter.py]: If you use some extra functions for this, add them as private method to a class 
and its name  should start with "__"

File: filters/Filter.py [hunk 6]
```diff
@@ -26,15 +100,148 @@ def gini_index(X, y):
...
+
+            for i in range(X.shape[1]):
+                cur_mi = MI(X[:, i].reshape(-1, 1), y)
+                if cur_mi > max_mi:
+                    feature_index = i
+                    max_mi = cur_mi
+
+            return feature_index
+
+        def _MID(A, B, y):
```
Comment (LastShekel) [inline filters/Filter.py]: same as above

File: setup.py [hunk 7]
```diff
@@ -5,5 +5,6 @@
     name='ITMO_FS',
     version='0.01',
     packages=find_packages(),
-    long_description=open(join(dirname(__file__), 'README.md')).read(), install_requires=['numpy', 'utils']
+    long_description=open(join(dirname(__file__), 'README.md')).read(), install_requires=['numpy', 'utils',
```
Comment (LastShekel) [inline setup.py]: No sklearn usage

File: filters/Filter.py [hunk 8]
```diff
@@ -414,13 +341,14 @@ def run(self, x, y, store_scores=False, feature_names=None):
         selected_features = self.cutting_rule(feature_scores)
         return x[:, selected_features]
 
+
 def test():
```
Comment (LastShekel) [inline filters/Filter.py]: No test needed here

File: test/filter test/filter_test.py [hunk 9]
```diff
@@ -49,15 +51,15 @@ def test_add_del(self):
...
-        print(wrapper.best_score)
+    #
+    # def test_backward_selection(self):
+    #     data, target = self.basehock['X'], self.basehock['Y']
+    #     lr = LogisticRegression()
+    #     wrapper = BackwardSelection(lr, 100, GLOB_MEASURE["GiniIndex"])
+    #     wrapper.fit(data[:, :200], target)
+    #     print(wrapper.best_score)
+    #     wrapper.fit(data[:, :300], target)
+    #     print(wrapper.best_score)
```
Comment (LastShekel) [inline test/filter test/filter_test.py]: Why it is commented?

File: filters/Filter.py [hunk 10]
```diff
@@ -320,10 +321,13 @@ def __select_k(cls, scores, k, reverse=False):
 
 class Filter(object):
     def __init__(self, measure, cutting_rule):
-        try:
-            self.measure = GLOB_MEASURE[measure]
-        except KeyError:
-            raise KeyError("No %r measure yet" % measure)
+        if measure is str:
```
Comment (LastShekel) [inline filters/Filter.py]: type(measure)==str

Comment (Paul1298) [reply filters/Filter.py]: thanks!

File: filters/Filter.py (hunk 8 above)
Comment (Paul1298) [reply filters/Filter.py]: removed

File: filters/Filter.py [hunk 11]
```diff
@@ -166,7 +177,60 @@ def cal_entropy(y):
...
+                    given, the dtype will be float.
+                """
+            classes, class_idx = np.unique(labels_true, return_inverse=True)
+            clusters, cluster_idx = np.unique(labels_pred, return_inverse=True)
+            n_classes = classes.shape[0]
+            n_clusters = clusters.shape[0]
+            # Using coo_matrix to accelerate simple histogram calculation,
+            # i.e. bins are consecutive integers
+            # Currently, coo_matrix is faster than histogram2d for simple cases
+            contingency = sp.coo_matrix((np.ones(class_idx.shape[0]),
```
[the remark this answers is not in the document]
Comment (Paul1298) [reply filters/Filter.py]: done

File: filters/Filter.py [hunk 12]
```diff
@@ -83,7 +94,7 @@ def __calculate_F_ratio__(row, y_data):
             f_ratios.append(f_ratio)
         f_ratios = np.array(f_ratios)
         # return top n f_ratios TODO maybe add n into parameters?
-        n = 10
+        n = 3
```
[the remark this answers is not in the document]
Comment (Paul1298) [reply filters/Filter.py]: done

**Pull request merged**